File: flutter/packages/workout_log_core/lib/src/cycles/cycle_generator.dart
import '../models/block.dart';
import '../models/cycle.dart';
import '../models/session.dart';
import '../models/work_set.dart';

class CycleGeneratorException implements Exception {
  CycleGeneratorException(this.message);

  final String message;

  @override
  String toString() => message;
}

/// Expands a cycle template onto the real calendar, producing one schema-valid
/// session stub per template session. The planning-only `role` and `sets_reps`
/// fields are consumed here and never reach the output.
abstract final class CycleGenerator {
  static const weekdayAbbreviations = [
    'Mon',
    'Tue',
    'Wed',
    'Thu',
    'Fri',
    'Sat',
    'Sun',
  ];

  /// The first `count` dates on/after `start` whose weekday is a training day.
  static List<DateTime> trainingDates(
    DateTime start,
    List<String> trainingDays,
    int count,
  ) {
    final valid = trainingDays.where(weekdayAbbreviations.contains).toSet();
    if (valid.isEmpty && count > 0) {
      throw CycleGeneratorException(
        'no recognised training days in $trainingDays '
        '(expected any of $weekdayAbbreviations)',
      );
    }
    final out = <DateTime>[];
    var day = DateTime(start.year, start.month, start.day);
    while (out.length < count) {
      if (valid.contains(weekdayAbbreviations[day.weekday - 1])) out.add(day);
      // Stepping via the constructor rather than `Duration(days: 1)` keeps this
      // on calendar days: adding 24h across a DST boundary lands on the wrong one.
      day = DateTime(day.year, day.month, day.day + 1);
    }
    return out;
  }

  static List<Session> generate(Cycle cycle) {
    final start = DateTime.parse(cycle.startDate);
    final dates = trainingDates(
      start,
      cycle.trainingDays,
      cycle.sessions.length,
    );
    return [
      for (var i = 0; i < cycle.sessions.length; i++)
        sessionFromTemplate(cycle.sessions[i], dates[i]),
    ];
  }

  /// `enforceWeekday` is what separates the two callers: expanding a whole cycle
  /// must fail loudly when the calendar has drifted, but a person creating one
  /// session on a day that suits them is not an error.
  static Session sessionFromTemplate(
    CycleSession template,
    DateTime when, {
    bool enforceWeekday = true,
  }) {
    final actual = weekdayAbbreviations[when.weekday - 1];
    final expected = template.weekday;
    if (enforceWeekday && expected != null && expected != actual) {
      throw CycleGeneratorException(
        '${template.cycleDay}: calendar says $actual but template says $expected',
      );
    }
    return Session(
      date: isoDate(when),
      cycleDay: template.cycleDay,
      notes: (template.sessionNotes?.isEmpty ?? true)
          ? null
          : template.sessionNotes,
      blocks: [for (final b in template.blocks) blockFromTemplate(b)],
    );
  }

  static Block blockFromTemplate(BlockTemplate template) {
    final notes =
        (template.notes?.isEmpty ?? true) ? null : template.notes;
    switch (template.type) {
      case 'cardio':
        return CardioBlock(
          machine: template.machine ?? '',
          durationMin: template.durationMin,
        );
      case 'strength':
        return StrengthBlock(
          exercise: template.exercise!,
          sets: [for (final reps in template.setsReps) WorkSet(reps: reps)],
          notes: notes,
        );
      case 'metcon':
        return MetconBlock(
          format: template.format,
          scheme: template.scheme,
          exercises: [for (final e in template.exercises) e],
          notes: notes,
        );
      case 'cooldown':
        return CooldownBlock();
      default:
        throw CycleGeneratorException(
          'unknown block type: "${template.type}"',
        );
    }
  }

  static String isoDate(DateTime when) =>
      '${when.year.toString().padLeft(4, '0')}-'
      '${when.month.toString().padLeft(2, '0')}-'
      '${when.day.toString().padLeft(2, '0')}';
}

File: crates/workout-log-core/src/cycles/cycle_generator.rs
//! Expands a cycle definition onto the calendar and into session stubs.
//!
//! Planning-only fields — `role`, `sets_reps`, `weekday`, `title`, `week` —
//! are consumed here and never reach a session file.

use chrono::{Datelike, Days, NaiveDate};

use super::cycle_day::CycleDay;
use crate::models::block::{Block, CardioBlock, CooldownBlock, StrengthBlock};
use crate::models::cycle::{BlockTemplate, Cycle, CycleSession};
use crate::models::metcon::MetconBlock;
use crate::models::session::{Kind, Session};
use crate::models::work_set::WorkSet;

#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
#[error("{0}")]
pub struct CycleGeneratorError(pub String);

type Result<T> = std::result::Result<T, CycleGeneratorError>;

pub const WEEKDAY_ABBREVIATIONS: [&str; 7] = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"];

fn abbreviation(date: NaiveDate) -> &'static str {
    WEEKDAY_ABBREVIATIONS[date.weekday().num_days_from_monday() as usize]
}

/// Walks the calendar from `start` inclusive, collecting dates whose weekday is
/// in `training_days`, until `count` are found.
pub fn training_dates(
    start: NaiveDate,
    training_days: &[String],
    count: usize,
) -> Result<Vec<NaiveDate>> {
    let wanted: Vec<&str> = WEEKDAY_ABBREVIATIONS
        .into_iter()
        .filter(|abbrev| training_days.iter().any(|d| d == abbrev))
        .collect();
    if wanted.is_empty() {
        if count > 0 {
            return Err(CycleGeneratorError(format!(
                "no recognised training days in {} (expected any of {})",
                training_days.join(", "),
                WEEKDAY_ABBREVIATIONS.join(", ")
            )));
        }
        return Ok(Vec::new());
    }
    let mut dates = Vec::with_capacity(count);
    let mut cursor = start;
    while dates.len() < count {
        if wanted.contains(&abbreviation(cursor)) {
            dates.push(cursor);
        }
        // Calendar-day stepping, so a DST boundary cannot shift the walk.
        cursor = cursor
            .checked_add_days(Days::new(1))
            .ok_or_else(|| CycleGeneratorError("ran off the end of the calendar".to_string()))?;
    }
    Ok(dates)
}

pub fn generate(cycle: &Cycle) -> Result<Vec<Session>> {
    let start = NaiveDate::parse_from_str(&cycle.start_date, "%Y-%m-%d")
        .map_err(|_| CycleGeneratorError(format!("\"{}\" is not an ISO date", cycle.start_date)))?;
    let dates = training_dates(start, &cycle.training_days, cycle.sessions.len())?;
    cycle
        .sessions
        .iter()
        .zip(dates)
        .map(|(template, date)| session_from_template(template, date, true))
        .collect()
}

pub fn session_from_template(
    template: &CycleSession,
    when: NaiveDate,
    enforce_weekday: bool,
) -> Result<Session> {
    if enforce_weekday && let Some(expected) = &template.weekday {
        let actual = abbreviation(when);
        if actual != expected {
            return Err(CycleGeneratorError(format!(
                "{}: calendar says {actual} but template says {expected}",
                template.cycle_day
            )));
        }
    }
    let blocks = template
        .blocks
        .iter()
        .map(block_from_template)
        .collect::<Result<Vec<_>>>()?;
    Ok(Session {
        date: iso_date(when),
        cycle_day: template.cycle_day.clone(),
        start_time: None,
        kind: Kind::Training,
        bodyweight_kg: None,
        notes: non_empty(template.session_notes.as_deref()),
        blocks,
    })
}

pub fn block_from_template(template: &BlockTemplate) -> Result<Block> {
    let notes = non_empty(template.notes.as_deref());
    Ok(match template.block_type.as_str() {
        "cardio" => Block::Cardio(CardioBlock {
            machine: template.machine.clone().unwrap_or_default(),
            duration_min: template.duration_min,
            distance_m: None,
            end_time: None,
        }),
        "strength" => {
            let exercise = non_empty(template.exercise.as_deref()).ok_or_else(|| {
                CycleGeneratorError("a strength block in this cycle has no exercise yet".into())
            })?;
            Block::Strength(StrengthBlock {
                exercise,
                sets: template
                    .sets_reps()
                    .iter()
                    .map(|&reps| WorkSet {
                        reps: Some(reps),
                        ..WorkSet::default()
                    })
                    .collect(),
                start_time: None,
                end_time: None,
                notes,
            })
        }
        "metcon" => Block::Metcon(MetconBlock {
            format: template.format,
            scheme: template.scheme.clone(),
            exercises: template.exercises().to_vec(),
            rounds: None,
            start_time: None,
            end_time: None,
            notes,
        }),
        "cooldown" => Block::Cooldown(CooldownBlock::default()),
        other => {
            return Err(CycleGeneratorError(format!(
                "unknown block type: \"{other}\""
            )));
        }
    })
}

/// A planned session as the muscle map should see it: blocks that cannot be
/// built yet are skipped, and a named-but-setless lift still scores one slot.
pub fn preview(template: &CycleSession, on: Option<NaiveDate>) -> Session {
    let when = on.unwrap_or_else(|| NaiveDate::from_ymd_opt(2000, 1, 1).expect("valid date"));
    let blocks = template
        .blocks
        .iter()
        .filter_map(|b| block_from_template(b).ok())
        .map(|block| match block {
            Block::Strength(mut strength) if strength.sets.is_empty() => {
                strength.sets.push(WorkSet::default());
                Block::Strength(strength)
            }
            other => other,
        })
        .collect();
    Session {
        date: iso_date(when),
        cycle_day: template.cycle_day.clone(),
        start_time: None,
        kind: Kind::Training,
        bodyweight_kg: None,
        notes: None,
        blocks,
    }
}

pub fn iso_date(when: NaiveDate) -> String {
    format!("{:04}-{:02}-{:02}", when.year(), when.month(), when.day())
}

pub fn cycle_day_of(template: &CycleSession) -> Option<CycleDay> {
    CycleDay::try_parse(&template.cycle_day)
}

fn non_empty(text: Option<&str>) -> Option<String> {
    text.filter(|t| !t.is_empty()).map(str::to_string)
}
